core/src/plugins/kubernetes/results-cache.ts

```diff
@@ -37,28 +60,111 @@ export interface StoreResultParams<A extends RunAction | TestAction, R extends C
 
 export function composeCacheableResult({
   result,
-  action,
   namespaceStatus,
 }: {
   result: RunResult
-  action: Action
   namespaceStatus: NamespaceStatus
 }): CacheableResult {
   return {
     ...result,
     namespaceStatus,
-    actionName: action.name,
   }
 }
 
 export function toActionStatus<T extends CacheableResult>(detail: T): ActionStatus {
   return { state: runResultToActionState(detail), detail, outputs: { log: detail.log } }
 }
 
-export interface ResultCache<A extends RunAction | TestAction, R extends CacheableResult> {
+export interface ResultCache<A extends CacheableAction, R extends CacheableResult> {
   load(params: LoadResultParams<A>): Promise<R | undefined>
 
-  store(params: StoreResultParams<A, R>): Promise<R>
+  store(params: StoreResultParams<A, R>): Promise<R | undefined>
 
   clear(param: ClearResultParams<A>): Promise<void>
 }
+
+export type SchemaVersion = `v${number}`
+
+/**
+ * Increment the current result schema format version
+ * if the result format changes breaking backwards-compatibility,
+ * e.g. serialization format.
+ */
+export const currentResultSchemaVersion: SchemaVersion = "v1"
+
+export type CreateCacheableResultKeyParams = {
+  schemaVersion: SchemaVersion
+  ctx: PluginContext
+  action: CacheableAction
+}
+
+export class StructuredCacheKey {
+  private readonly schemaVersion: SchemaVersion
+  private readonly projectName: string
+  private readonly actionKind: string
+  private readonly actionName: string
+  private readonly actionType: string
+  private readonly actionVersion: string
+
+  constructor({ schemaVersion, ctx, action }: CreateCacheableResultKeyParams) {
+    this.schemaVersion = schemaVersion
+    this.projectName = ctx.projectName
+    this.actionKind = action.kind
+    this.actionName = action.name
+    this.actionType = action.type
+    this.actionVersion = action.versionString()
+  }
+
+  @Memoize()
+  public calculate(): string {
+    const key = `${this.projectName}--${this.actionType}.${this.actionName}--${this.actionVersion}--${this.schemaVersion}`
```

Comment:
for the team cache, it will be the concern of the backend to distinguish cache entries from different schema versions. Let's implement the file cache the same way.
```suggestion
    const key = `${this.projectName}--${this.actionType}.${this.actionName}--${this.actionVersion}`
```

Reply:
Shouldn't we always avoid cache key collisions? Backend can do whatever it needs, but we decided that cache key construction algorithm will be controlled by the client. Also, cache key must be unique, otherwise we'll get collisions.

So, backend will store the schema version in DB as a separate column, it's good and convenient for its needs.
But, if we remove the schema version from the cache key calculation, then 2 result objects of 2 different shapes will result into the same cache key.

Thus, the unique identifier will be a tuple of `<cacheKey, schemaVersion>`. It will require a composite unique index in DB, and will make the FS-based implementation a bit more complex. We'll still have to append the schema version to calculate the unique key for the local cache.

Why don't we just use the current key format? Backend can still maintain its own data structures and validation, but it doesn't look like the backend internal implementation should affect the way the client builds the cache keys.

Reply:
Ah, I see, thanks for elaborating this!

Actually, in that case we agree to have a kind of isolated logical buckets for different schema versions.
That makes sense. So the requirement is just to ensure cache key uniqueness in the scope of a single schema version. In that case we also avoid cache key format change on every change of the data shape.

I'll update that part.

Reply:
Done in fec1de1792369c23daa1ea7c00df53e9460565a6.